src/Zio/UPath.cs

```diff
@@ -296,165 +277,159 @@ internal static StringBuilder GetSharedStringBuilder()
+            //}
 
-                if (c == DirectorySeparator || c == '\\')
-                {
-                    // optimization: If we don't expect to process the path
-                    // and we only have a trailing / or \\, then just perform
-                    // a substring on the path
-                    if (!processParts && i + 1 == path.Length)
-                        return path.Substring(0, path.Length - 1);
+            if (c == '.')
+            {
+                dotCount++;
+            }
+            else if (c == DirectorySeparator || c == '\\')
+            {
+                // optimization: If we don't expect to process the path
+                // and we only have a trailing / or \\, then just perform
+                // a substring on the path
+                if (!processParts && i + 1 == path.Length)
+                    return path.Substring(0, path.Length - 1);
 
-                    if (c == '\\')
-                        processParts = true;
+                if (c == '\\')
+                    processParts = true;
 
-                    var endIndex = i - 1;
-                    for (i++; i < path.Length; i++)
+                var endIndex = i - 1;
+                for (i++; i < path.Length; i++)
+                {
+                    c = path[i];
+                    if (c == DirectorySeparator || c == '\\')
                     {
-                        c = path[i];
-                        if (c == DirectorySeparator || c == '\\')
-                        {
-                            // If we have consecutive / or \\, we need to process parts
-                            processParts = true;
-                            continue;
-                        }
-                        break;
+                        // If we have consecutive / or \\, we need to process parts
+                        processParts = true;
+                        continue;
                     }
+                    break;
+                }
 
-                    if (endIndex >= lastIndex || endIndex == -1)
-                    {
-                        var part = new TextSlice(lastIndex, endIndex);
-                        parts.Add(part);
+                if (endIndex >= lastIndex || endIndex == -1)
+                {
+                    var part = new TextSlice(lastIndex, endIndex);
+                    parts.Add(part);
 
-                        // If the previous part had only dots, we need to process it
-                        if (part.Length == dotCount)
-                            processParts = true;
-                    }
-                    dotCount = c == '.' ? 1 : 0;
-                    lastIndex = i;
+                    if (dotCount > 0 &&                     // has dots
+                        dotCount == part.Length &&          // only dots
+                        (dotCount != 2 || parts.Count > 1)) // Skip ".." if it's the first part
```

Comment:
I made that logic a bit more complex, to only really set processParts, when really needed, so more paths are using the fast path only.